server/src/main/java/org/opensearch/action/search/SearchRequestContext.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.search;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.lucene.search.TotalHits;
+
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class holds request-level context for search queries at the coordinator node
+ *
+ * @opensearch.internal
+ */
+public class SearchRequestContext {
+    private final SearchRequestOperationsListener searchRequestOperationsListener;
+    private long absoluteStartNanos;
+    private final Map<String, Long> phaseTookMap;
+    private TotalHits totalHits;
+    private final EnumMap<ShardStatsFieldNames, Integer> shardStats;
+
+    public SearchRequestContext() {
+        this(new SearchRequestOperationsListener.CompositeListener(List.of(), LogManager.getLogger()));
+    }
+
+    public SearchRequestContext(SearchRequestOperationsListener searchRequestOperationsListener) {
+        this.searchRequestOperationsListener = searchRequestOperationsListener;
+        this.absoluteStartNanos = System.nanoTime();
+        this.phaseTookMap = new HashMap<>();
+        this.shardStats = new EnumMap<>(ShardStatsFieldNames.class);
+    }
+
+    SearchRequestOperationsListener getSearchRequestOperationsListener() {
+        return searchRequestOperationsListener;
+    }
+
+    void updatePhaseTookMap(String phaseName, Long tookTime) {
+        this.phaseTookMap.put(phaseName, tookTime);
+    }
+
+    Map<String, Long> phaseTookMap() {
+        return phaseTookMap;
+    }
+
+    /**
+     * Override absoluteStartNanos set in constructor.
+     * For testing only
+     */
+    void setAbsoluteStartNanos(long absoluteStartNanos) {
+        this.absoluteStartNanos = absoluteStartNanos;
+    }
+
+    /**
+     * Request start time in nanos
+     */
+    long getAbsoluteStartNanos() {
+        return absoluteStartNanos;
+    }
+
+    void setTotalHits(TotalHits totalHits) {
+        this.totalHits = totalHits;
+    }
+
+    TotalHits totalHits() {
+        return totalHits;
+    }
+
+    void setShardStats(int total, int successful, int skipped, int failed) {
+        this.shardStats.put(ShardStatsFieldNames.SEARCH_REQUEST_SLOWLOG_SHARD_TOTAL, total);
+        this.shardStats.put(ShardStatsFieldNames.SEARCH_REQUEST_SLOWLOG_SHARD_SUCCESSFUL, successful);
+        this.shardStats.put(ShardStatsFieldNames.SEARCH_REQUEST_SLOWLOG_SHARD_SKIPPED, skipped);
+        this.shardStats.put(ShardStatsFieldNames.SEARCH_REQUEST_SLOWLOG_SHARD_FAILED, failed);
+    }
+
+    String formattedShardStats() {
+        if (shardStats.isEmpty()) {
+            return "";
+        } else {
+            return String.format(
```

Comment:
ForbiddenAPIs is complaining. The first argument to every call to `String.format()` should be `Locale.ROOT`, to avoid using the system locale.